irods/manager/data_object_manager.py

```diff
@@ -317,29 +322,92 @@ def open(self, path, mode, create = True, finalize_on_close = True, **options):
-        )
-        message = iRODSMessage('RODS_API_REQ', msg=message_body,
-                               int_info=api_number['DATA_OBJ_OPEN_AN'])
+        def make_FileOpenRequest():
+            return  FileOpenRequest(
+                objPath=path,
+                createMode=0,
+                openFlags=flags,
+                offset=0,
+                dataSize=-1,
+                numThreads=self.sess.numThreads,
+                oprType=oprType,
+                KeyValPair_PI=StringStringMap(options),
+            )
+
+        # preserve RESC HIER but do not pass through GET_HOST_FOR_* call
+        requested_hierarchy = options.pop(kw.RESC_HIER_STR_KW, None)
+        message_body = make_FileOpenRequest()
 
         conn = self.sess.pool.get_connection()
+        redirected_host = ''
+
+        use_get_rescinfo_apis = False
+
+        if allow_redirect and conn.server_version >= (4,3,1):
+            choices = ('PUT','GET')
+            key_val = [(k,v) for k,v in returned_values.items() if k in choices]
+            if len(key_val) != 1:
+                raise ValueError("If provided, return_host argument must have 1 key in {}".format(choices))
+            (key, _) = key_val[0]
+            api_index = "GET_RESCINFO_FOR_{}_AN".format(key)
+            message = iRODSMessage('RODS_API_REQ', msg=message_body,
+                                   int_info=api_number[api_index])
+            conn.send(message)
+            response = conn.recv()
+            msg = response.get_main_message( STR_PI )
+            use_get_rescinfo_apis = True
+
+            # Get the information needed for the redirect
+            _ = json.loads(msg.myStr)
+            redirected_host = _["host"]
+            requested_hierarchy = _["resc_hier"]
+
+            if redirected_host:
+                returned_values[key] = redirected_host
+
+        target_zone = list(filter(None, path.split('/')))
+        if target_zone:
+            target_zone = target_zone[0]
+
+        directed_sess = self.sess
+
+        if redirected_host and use_get_rescinfo_apis:
+            # Redirect only if the local zone is being targeted, and if the hostname is changed from the original.
+            if target_zone == self.sess.zone and (self.sess.host != redirected_host 
+                                                  or str(allow_redirect).lower() == 'force' # TODO - remove - for debug only
+                                                  ):
+                # This is the actual redirect.
+                directed_sess = self.sess.clone(host = redirected_host)
+                returned_values['session'] = directed_sess
+                conn = directed_sess.pool.get_connection()
+                logger.info('redirect_to_host = %s', redirected_host)
+
+        # Restore RESC HIER for DATA_OBJ_OPEN call
+        if requested_hierarchy is not None:
+            options[kw.RESC_HIER_STR_KW] = requested_hierarchy
+            message_body = make_FileOpenRequest()
```

Comment:
Should this be part of a try/finally block? I see one explicit `raise` above, but I'm sure others can `raise` as well.

Reply:
Believe the aforementioned raise will probably disappear, as it was just a debug remnant.

Reply:
The data object open( ) routine's logic has been sufficiently changed since @alanking made this comment, and the `raise` statement is now gone, so I think we're good and no need to clean up after failures.